db/db.ts

```diff
@@ -448,3 +453,103 @@ export const getNonGrapherExplorerViewCount = (
             AND grapherId IS NULL`
     ).then((res) => res?.count ?? 0)
 }
+
+/**
+ * 1. Fetch all records in tag_graph, isTopic = true when there is a published TP/LTP/Article with the same slug as the tag
+ * 2. Group tags by their parentId
+ * 3. Return the flat tag graph along with a __rootId property so that the UI knows which record is the root node
+ */
+export async function getFlatTagGraph(knex: KnexReadonlyTransaction): Promise<
+    FlatTagGraph & {
+        __rootId: number
+    }
+> {
+    const tagGraphByParentId = await knexRaw<FlatTagGraphNode>(
+        knex,
+        `-- sql
+        SELECT
+            tg.parentId,
+            tg.childId,
+            tg.weight,
+            t.name,
+            t.slug,
+            IFNULL((p.type IN (:types) AND p.published = 1 AND t.slug IS NOT NULL), FALSE) AS isTopic
+        FROM
+            tag_graph tg
+        LEFT JOIN tags t ON
+            tg.childId = t.id
+        LEFT JOIN posts_gdocs p on
+            t.slug = p.slug
+        -- order by descending weight, tiebreak by name
+        ORDER BY tg.weight DESC, t.name ASC`,
+        {
+            types: [
+                OwidGdocType.TopicPage,
+                OwidGdocType.LinearTopicPage,
+                // For sub-topics e.g. Nuclear Energy we use the article format
+                OwidGdocType.Article,
+            ],
+        }
+    ).then((rows) => groupBy(rows, "parentId"))
+
+    const tagGraphRootIdResult = await knexRawFirst<{
+        id: number
+    }>(
+        knex,
+        `-- sql
+        SELECT id FROM tags WHERE name = "${TagGraphRootName}"`
+    )
+    if (!tagGraphRootIdResult) throw new Error("Tag graph root not found")
+
+    return { ...tagGraphByParentId, __rootId: tagGraphRootIdResult.id }
+}
+
+export async function updateTagGraph(
+    knex: KnexReadWriteTransaction,
+    tagGraph: FlatTagGraph
+): Promise<void> {
+    const tagGraphRows: {
+        parentId: number
+        childId: number
+        weight: number
+    }[] = []
+
+    for (const children of Object.values(tagGraph)) {
+        for (const child of children) {
+            tagGraphRows.push({
+                parentId: child.parentId,
+                childId: child.childId,
+                weight: child.weight,
+            })
+        }
+    }
+
+    await knex("tag_graph").delete()
+    await knex("tag_graph").insert(tagGraphRows)
```

Comment:
This is an easy way to do the update but then the updatedAt timestamp will not be useful information. I think being able to know which tags have been updated might be worth the slight complication to do a select all and then do the add/update/delete operations accordingly. What do you think?

Reply:
The tag graph doesn't have `updatedAt` - do you think it should?

Reply:
I think it can be really useful - e.g. when something goes wrong and you want to undo some change from yesterday or so it's really nice if that query can identify the three rows that actually changed instead of having to fetch all from a backup.

That said I'm not sure about whether it's worth changing this PR for it.